noir-r1cs/src/noir_witness.rs

```diff
@@ -139,3 +140,46 @@ fn encode_input(
     }
     Ok(())
 }
+
+/// Trait which is used to add witness RNG for IOPattern
+pub trait WitnessIOPattern {
+    /// Schedule absorption of circuit shape (2 scalars): (num_constraints,
+    /// num_witnesses).
+    fn add_shape(self) -> Self;
+
+    /// Schedule absorption of `num_pub_inputs` public input scalars.
+    fn add_public_inputs(self, num_pub_inputs: usize) -> Self;
+
+    /// Schedule absorption of `num_challenges` Fiat–Shamir challenges for
+    /// LogUp/Spice.
+    fn add_logup_challenges(self, num_challenges: usize) -> Self;
+}
+
+impl<IOPattern> WitnessIOPattern for IOPattern
+where
+    IOPattern: FieldDomainSeparator<FieldElement>,
+{
+    fn add_shape(self) -> Self {
+        self.add_scalars(2, "shape")
+    }
+
+    fn add_public_inputs(self, num_pub_inputs: usize) -> Self {
+        if num_pub_inputs > 0 {
+            self.add_scalars(num_pub_inputs, "pub_inputs")
+        } else {
+            self
+        }
+    }
+
+    fn add_logup_challenges(self, num_challenges: usize) -> Self {
+        if num_challenges > 0 {
+            FieldDomainSeparator::<FieldElement>::challenge_scalars(
```

Comment:
use `fill_challenge_scalars`

Reply:
using `fill_logup_challenges`, it needs to be the same name as above